File: src/2025-02-24-deepclaude-rust/cpp/exercises/practice/zebra-puzzle/zebra_puzzle.cpp
#include "zebra_puzzle.h"
#include <array>
#include <vector>
#include <algorithm>

namespace zebra_puzzle {

Solution Solve() {
    // Define house positions (1-5)
    const int FIRST = 0;
    const int SECOND = 1;
    const int MIDDLE = 2;
    const int FOURTH = 3;
    const int FIFTH = 4;
    
    // Define attributes
    enum Color { RED, GREEN, IVORY, YELLOW, BLUE };
    enum Nationality { ENGLISH, SPANISH, UKRAINIAN, NORWEGIAN, JAPANESE };
    enum Pet { DOG, SNAIL, FOX, HORSE, ZEBRA };
    enum Drink { COFFEE, TEA, MILK, JUICE, WATER };
    enum Hobby { PAINTER, DANCING, CHESS, READING, FOOTBALL };
    
    // Create arrays to hold the final house configuration
    std::array<Color, 5> colors;
    std::array<Nationality, 5> nationalities;
    std::array<Pet, 5> pets;
    std::array<Drink, 5> drinks;
    std::array<Hobby, 5> hobbies;
    
    // Generate all possible permutations for testing
    std::vector<Color> colorPerm = {RED, GREEN, IVORY, YELLOW, BLUE};
    std::vector<Nationality> nationalityPerm = {ENGLISH, SPANISH, UKRAINIAN, NORWEGIAN, JAPANESE};
    std::vector<Pet> petPerm = {DOG, SNAIL, FOX, HORSE, ZEBRA};
    std::vector<Drink> drinkPerm = {COFFEE, TEA, MILK, JUICE, WATER};
    std::vector<Hobby> hobbyPerm = {PAINTER, DANCING, CHESS, READING, FOOTBALL};
    
    // Try all permutations to find a solution that satisfies all constraints
    do {
        colors = std::array<Color, 5>(colorPerm.begin(), colorPerm.end());
        
        do {
            nationalities = std::array<Nationality, 5>(nationalityPerm.begin(), nationalityPerm.end());
            
            // Constraint: The Norwegian lives in the first house
            if (nationalities[FIRST] != NORWEGIAN) continue;
            
            // Constraint: The Englishman lives in the red house
            if (nationalities[std::find(colors.begin(), colors.end(), RED) - colors.begin()] != ENGLISH) continue;
            
            // Constraint: The Norwegian lives next to the blue house
            if (!(colors[SECOND] == BLUE)) continue;
            
            do {
                pets = std::array<Pet, 5>(petPerm.begin(), petPerm.end());
                
                // Constraint: The Spaniard owns the dog
                if (pets[std::find(nationalities.begin(), nationalities.end(), SPANISH) - nationalities.begin()] != DOG) continue;
                
                do {
                    drinks = std::array<Drink, 5>(drinkPerm.begin(), drinkPerm.end());
                    
                    // Constraint: The person in the green house drinks coffee
                    if (drinks[std::find(colors.begin(), colors.end(), GREEN) - colors.begin()] != COFFEE) continue;
                    
                    // Constraint: The Ukrainian drinks tea
                    if (drinks[std::find(nationalities.begin(), nationalities.end(), UKRAINIAN) - nationalities.begin()] != TEA) continue;
                    
                    // Constraint: The person in the middle house drinks milk
                    if (drinks[MIDDLE] != MILK) continue;
                    
                    do {
                        hobbies = std::array<Hobby, 5>(hobbyPerm.begin(), hobbyPerm.end());
                        
                        // Constraint: The snail owner likes to go dancing
                        if (hobbies[std::find(pets.begin(), pets.end(), SNAIL) - pets.begin()] != DANCING) continue;
                        
                        // Constraint: The person in the yellow house is a painter
                        if (hobbies[std::find(colors.begin(), colors.end(), YELLOW) - colors.begin()] != PAINTER) continue;
                        
                        // Constraint: The person who plays football drinks orange juice
                        if (drinks[std::find(hobbies.begin(), hobbies.end(), FOOTBALL) - hobbies.begin()] != JUICE) continue;
                        
                        // Constraint: The Japanese person plays chess
                        if (hobbies[std::find(nationalities.begin(), nationalities.end(), JAPANESE) - nationalities.begin()] != CHESS) continue;
                        
                        // Constraint: The green house is immediately to the right of the ivory house
                        bool greenRightOfIvory = false;
                        for (int i = 0; i < 4; i++) {
                            if (colors[i] == IVORY && colors[i+1] == GREEN) {
                                greenRightOfIvory = true;
                                break;
                            }
                        }
                        if (!greenRightOfIvory) continue;
                        
                        // Constraint: The person who enjoys reading lives in the house next to the person with the fox
                        int foxPos = std::find(pets.begin(), pets.end(), FOX) - pets.begin();
                        int readingPos = std::find(hobbies.begin(), hobbies.end(), READING) - hobbies.begin();
                        bool readerNextToFox = abs(foxPos - readingPos) == 1;
                        if (!readerNextToFox) continue;
                        
                        // Constraint: The painter's house is next to the house with the horse
                        int painterPos = std::find(hobbies.begin(), hobbies.end(), PAINTER) - hobbies.begin();
                        int horsePos = std::find(pets.begin(), pets.end(), HORSE) - pets.begin();
                        bool painterNextToHorse = abs(painterPos - horsePos) == 1;
                        if (!painterNextToHorse) continue;
                        
                        // If we get here, all constraints are satisfied - we found a solution!
                        std::string drinksWater;
                        std::string ownsZebra;
                        
                        int waterDrinker = std::find(drinks.begin(), drinks.end(), WATER) - drinks.begin();
                        int zebraOwner = std::find(pets.begin(), pets.end(), ZEBRA) - pets.begin();
                        
                        switch (nationalities[waterDrinker]) {
                            case ENGLISH: drinksWater = "Englishman"; break;
                            case SPANISH: drinksWater = "Spaniard"; break;
                            case UKRAINIAN: drinksWater = "Ukrainian"; break;
                            case NORWEGIAN: drinksWater = "Norwegian"; break;
                            case JAPANESE: drinksWater = "Japanese"; break;
                        }
                        
                        switch (nationalities[zebraOwner]) {
                            case ENGLISH: ownsZebra = "Englishman"; break;
                            case SPANISH: ownsZebra = "Spaniard"; break;
                            case UKRAINIAN: ownsZebra = "Ukrainian"; break;
                            case NORWEGIAN: ownsZebra = "Norwegian"; break;
                            case JAPANESE: ownsZebra = "Japanese"; break;
                        }
                        
                        return Solution{drinksWater, ownsZebra};
                    } while (std::next_permutation(hobbyPerm.begin(), hobbyPerm.end()));
                } while (std::next_permutation(drinkPerm.begin(), drinkPerm.end()));
            } while (std::next_permutation(petPerm.begin(), petPerm.end()));
        } while (std::next_permutation(nationalityPerm.begin(), nationalityPerm.end()));
    } while (std::next_permutation(colorPerm.begin(), colorPerm.end()));
    
    // Should never reach here if the puzzle has a unique solution
    return Solution{"", ""};
}

}  // namespace zebra_puzzle

File: src/2025-03-16-12-54-49--a-deepclaude-full-architect-python-0316-2/python/exercises/practice/zebra-puzzle/zebra_puzzle.py
import itertools

def solution():
    """Solve the zebra puzzle and return the solution."""
    # Define houses by position (0-indexed)
    houses = range(5)
    
    # Generate all possible arrangements
    for colors in itertools.permutations(['red', 'green', 'ivory', 'yellow', 'blue']):
        # Constraint 6: The green house is immediately to the right of the ivory house
        if colors.index('green') != colors.index('ivory') + 1:
            continue
            
        for nationalities in itertools.permutations(['Englishman', 'Spaniard', 'Ukrainian', 'Norwegian', 'Japanese']):
            # Constraint 2: The Englishman lives in the red house
            if colors[nationalities.index('Englishman')] != 'red':
                continue
                
            # Constraint 10: The Norwegian lives in the first house
            if nationalities[0] != 'Norwegian':
                continue
                
            # Constraint 15: The Norwegian lives next to the blue house
            if abs(colors.index('blue') - nationalities.index('Norwegian')) != 1:
                continue
                
            for pets in itertools.permutations(['dog', 'snail', 'fox', 'horse', 'zebra']):
                # Constraint 3: The Spaniard owns the dog
                if pets[nationalities.index('Spaniard')] != 'dog':
                    continue
                    
                for drinks in itertools.permutations(['water', 'coffee', 'tea', 'milk', 'orange juice']):
                    # Constraint 4: The person in the green house drinks coffee
                    if drinks[colors.index('green')] != 'coffee':
                        continue
                        
                    # Constraint 5: The Ukrainian drinks tea
                    if drinks[nationalities.index('Ukrainian')] != 'tea':
                        continue
                        
                    # Constraint 9: The person in the middle house drinks milk
                    if drinks[2] != 'milk':
                        continue
                        
                    # Constraint 13: The person who plays football drinks orange juice
                    for hobbies in itertools.permutations(['painting', 'dancing', 'reading', 'football', 'chess']):
                        # Constraint 7: The snail owner likes to go dancing
                        if hobbies[pets.index('snail')] != 'dancing':
                            continue
                            
                        # Constraint 8: The person in the yellow house is a painter
                        if hobbies[colors.index('yellow')] != 'painting':
                            continue
                            
                        # Constraint 11: The person who enjoys reading lives in the house next to the person with the fox
                        if abs(hobbies.index('reading') - pets.index('fox')) != 1:
                            continue
                            
                        # Constraint 12: The painter's house is next to the house with the horse
                        if abs(hobbies.index('painting') - pets.index('horse')) != 1:
                            continue
                            
                        # Constraint 13: The person who plays football drinks orange juice
                        if drinks[hobbies.index('football')] != 'orange juice':
                            continue
                            
                        # Constraint 14: The Japanese person plays chess
                        if hobbies[nationalities.index('Japanese')] != 'chess':
                            continue
                        
                        # All constraints satisfied, we found the solution
                        water_drinker = nationalities[drinks.index('water')]
                        zebra_owner = nationalities[pets.index('zebra')]
                        
                        return water_drinker, zebra_owner
    
    # If no solution is found
    return None, None

def drinks_water():
    """Return the nationality of the person who drinks water."""
    return solution()[0]


def owns_zebra():
    """Return the nationality of the person who owns the zebra."""
    return solution()[1]
